crates/sui-synthetic-ingestion/src/synthetic_ingestion.rs

```diff
@@ -1,36 +1,64 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::SyntheticIngestionConfig;
 use simulacrum::Simulacrum;
+use std::path::PathBuf;
 use sui_test_transaction_builder::TestTransactionBuilder;
 use sui_types::crypto::get_account_key_pair;
 use sui_types::effects::TransactionEffectsAPI;
 use sui_types::gas_coin::MIST_PER_SUI;
 use sui_types::utils::to_sender_signed_transaction;
+use tokio::fs;
 use tracing::info;
 
+#[derive(clap::Parser, Debug, Clone)]
+pub struct Config {
+    /// Directory to write the ingestion data to.
+    #[clap(long)]
+    pub ingestion_dir: PathBuf,
+    /// Customize the first checkpoint sequence number in the workload.
+    /// This is useful if we want to generate workload to benchmark a non-empty database.
+    #[clap(long, default_value_t = 0)]
+    pub starting_checkpoint: u64,
+    /// Total number of synthetic checkpoints to generate.
+    #[clap(long, default_value_t = 2000)]
+    pub num_checkpoints: u64,
+    /// Number of transactions in a checkpoint.
+    #[clap(long, default_value_t = 200)]
+    pub checkpoint_size: u64,
+}
+
 // TODO: Simulacrum does serial execution which could be slow if
 // we need to generate a large number of transactions.
 // We may want to make Simulacrum support parallel execution.
 
-pub(crate) fn generate_ingestion(config: SyntheticIngestionConfig) {
+pub async fn generate_ingestion(config: Config) {
     info!("Generating synthetic ingestion data. config: {:?}", config);
     let timer = std::time::Instant::now();
     let mut sim = Simulacrum::new();
-    let SyntheticIngestionConfig {
+    let Config {
         ingestion_dir,
         checkpoint_size,
         num_checkpoints,
         starting_checkpoint,
     } = config;
-    sim.set_data_ingestion_path(ingestion_dir);
-    sim.override_last_checkpoint_number(starting_checkpoint - 1);
+    sim.set_data_ingestion_path(ingestion_dir.clone());
+    // Simulacrum will generate 0.chk as the genesis checkpoint.
+    // We do not need it and might even override if starting_checkpoint is 0.
+    fs::remove_file(ingestion_dir.join("0.chk")).await.unwrap();
 
     let gas_price = sim.reference_gas_price();
     let (sender, keypair) = get_account_key_pair();
-    let effects = sim.request_gas(sender, MIST_PER_SUI * 1000000).unwrap();
-    let mut gas_object = effects.created()[0].0;
+    let mut gas_object = {
+        let effects = sim.request_gas(sender, MIST_PER_SUI * 1000000).unwrap();
+        // `request_gas` will create a transaction, which we don't want to include in the benchmark.
+        // Put it in a checkpoint and then remove the checkpoint file.
+        sim.create_checkpoint();
+        fs::remove_file(ingestion_dir.join("1.chk")).await.unwrap();
```

Comment:
is `remove_file` necessary? as `std::fs::write(path.join(file_name), blob.to_bytes())?;` will overwrite `0.chk` and `1.chk` after `override_next_checkpoint_number`
- if the test range covers cp 0 and cp 1, it will overwrite
- otherwise these 2 files are not used?

Reply:
I was hoping to have a mode where when you run benchmark from an ingestion dir, it automatically scan through all files in it